File: Remote/Source/Remote Elements/Remote Element/Remote.m
//
// Remote.m
// Remote
//
// Created by Jason Cardwell on 6/1/11.
// Copyright (c) 2011 Moondeer Studios. All rights reserved.
//

#import "Remote.h"
#import "RemoteElement_Private.h"

static int       ddLogLevel   = LOG_LEVEL_WARN;
static const int msLogContext = LOG_CONTEXT_REMOTE;
#pragma unused(ddLogLevel, msLogContext)


@interface Remote ()
@property (nonatomic, strong, readwrite) NSDictionary * panels;
@end

@interface Remote (CoreDataGeneratedAccessors)
@property (nonatomic) NSDictionary * primitivePanels;
@end


@implementation Remote

@dynamic panels;

+ (REType)elementType { return RETypeRemote; }

- (void)setParentElement:(RemoteElement *)parentElement {}

- (RemoteElement *)parentElement { return nil; }

- (void)setTopBarHidden:(BOOL)topBarHidden {
  REOptions options = self.options;
  self.options = (topBarHidden
                  ? options | RERemoteOptionTopBarHidden
                  : options & ~RERemoteOptionTopBarHidden);
}

- (BOOL)isTopBarHidden {
  BOOL topBarHidden       = (self.options == RERemoteOptionTopBarHidden) ? YES : NO;
  BOOL faultyTopBarHidden = (self.options & RERemoteOptionTopBarHidden) ? YES : NO;
  assert(topBarHidden == faultyTopBarHidden);

  if (topBarHidden) assert(self.options == 1);

  return topBarHidden;
}

- (void)assignButtonGroup:(ButtonGroup *)buttonGroup assignment:(REPanelAssignment)assignment {
  NSMutableDictionary * panels = [self.panels mutableCopy];
  panels[@(assignment)]       = CollectionSafe(buttonGroup.uuid);
  self.panels                 = panels;
  buttonGroup.panelAssignment = assignment;
}

- (ButtonGroup *)buttonGroupForAssignment:(REPanelAssignment)assignment {
  NSString * uuid = NilSafe(self.panels[@(assignment)]);
  return (uuid ? [ButtonGroup existingObjectWithUUID:uuid context:self.managedObjectContext] : nil);
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Importing
////////////////////////////////////////////////////////////////////////////////


- (void)updateWithData:(NSDictionary *)data {
  /*
     {
      "uuid": "B0EA5B35-5CF6-40E9-B302-0F164D4A7ADD", // Home Screen
      "name": "Home Screen",
      "elementType": "remote",
      "key": "activity1",
      "options": "top-bar-hidden",
      "constraints": {
          "index": {
              "activityButtons": "F358CB82-496C-446D-8833-D4373777D23E",
              "bottomToolbar": "B81FFF61-4F56-43C1-9E60-3C399EB31C1B",
              "homeScreen": "B0EA5B35-5CF6-40E9-B302-0F164D4A7ADD" // Home Screen
          },
          "format": [
              "activityButtons.centerX = homeScreen.centerX",
              "activityButtons.centerY = homeScreen.centerY - 22",
              "bottomToolbar.bottom = homeScreen.bottom",
              "bottomToolbar.left = homeScreen.left",
              "bottomToolbar.right = homeScreen.right"
          ]
      },
      "backgroundColor": "black",
      "backgroundImage.uuid": "089D4A98-E7C1-472A-A0A3-30258BE42388", // Pro Dots.png
      "subelements": [ **ButtonGroup** ],
      "panels": {
          "left1": "7521F420-F677-44C9-97BA-4AF836779C21", // Left Overlay Panel
          "right1": "A6394F58-79C6-4B6C-962D-1DDD6BE1C36F", // Selection Panel
          "bottom1": "D505193C-C9D5-4D89-9C9F-D0B24EAE6D69", // DVR Activity Transport
          "top1": "A2EFC284-8F45-45B1-9A4B-DB448A2AEAE7" // DVR Activity Number Pad
      }
     }
   */

  [super updateWithData:data];

  NSDictionary * panels = data[@"panels"];

  if (panels && isDictionaryKind(panels)) {
    [panels enumerateKeysAndObjectsUsingBlock:
     ^(NSString * assignmentKey, NSString * uuid, BOOL * stop)
    {
      ButtonGroup * panel = (ButtonGroup *)memberOfCollectionWithUUID(self.subelements, uuid);

      if (panel) {
        REPanelAssignment assignment = panelAssignmentFromImportKey(assignmentKey);
        [self assignButtonGroup:panel assignment:assignment];
      }
    }];
  }

}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Exporting
////////////////////////////////////////////////////////////////////////////////

- (MSDictionary *)JSONDictionary {
  MSDictionary * dictionary = [super JSONDictionary];

  MSDictionary * panels = [MSDictionary dictionary];
  [self.panels enumerateKeysAndObjectsUsingBlock:^(NSNumber * key, NSString * uuid, BOOL *stop){
    NSString * k = panelKeyForPanelAssignment(IntValue(key));
    SafeSetValueForKey([self buttonGroupForAssignment:IntValue(key)].commentedUUID, k, panels);
  }];

  SafeSetValueForKey(panels, @"panels", dictionary);

  [dictionary compact];
  [dictionary compress];

  return dictionary;
}

@end

@implementation Remote (Debugging)

- (MSDictionary *)deepDescriptionDictionary {
  Remote * element = [self faultedObject];
  assert(element);

  MSDictionary * dd = [[super deepDescriptionDictionary] mutableCopy];
  dd[@"topBarHidden"] = BOOLString(element.topBarHidden);

  return (MSDictionary *)dd;
}

@end

File: Remote/Source/Remote Elements/Models/Remote.swift
//
//  Remote.swift
//  Remote
//
//  Created by Jason Cardwell on 11/16/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(Remote)
class Remote: RemoteElement {

  override var elementType: BaseType { return .Remote }

  @NSManaged var topBarHidden: Bool
  @NSManaged var activity: Activity?

  @NSManaged var panels: NSDictionary
  var panelAssignments: [NSNumber:String] {
    get { return panels as? [NSNumber:String] ?? [:] }
    set { panels = newValue }
  }

  override var parentElement: RemoteElement? { get { return nil } set {} }

  /**
  initWithPreset:

  :param: preset Preset
  */
  override init(preset: Preset) {
    super.init(preset: preset)

    topBarHidden = preset.topBarHidden ?? false
  }

  required init(context: NSManagedObjectContext, insert: Bool) {
      fatalError("init(context:insert:) has not been implemented")
  }

  /**
  initWithEntity:insertIntoManagedObjectContext:

  :param: entity NSEntityDescription
  :param: context NSManagedObjectContext?
  */
//  override init(entity: NSEntityDescription, insertIntoManagedObjectContext context: NSManagedObjectContext?) {
//    super.init(entity: entity, insertIntoManagedObjectContext: context)
//  }

  /**
  initWithContext:

  :param: context NSManagedObjectContext
  */
//  override init(context: NSManagedObjectContext) {
//    super.init(context: context)
//  }

  /**
  setButtonGroup:forPanelAssignment:

  :param: buttonGroup ButtonGroup?
  :param: assignment ButtonGroup.PanelAssignment
  */
  func setButtonGroup(buttonGroup: ButtonGroup?, forPanelAssignment assignment: ButtonGroup.PanelAssignment) {
    var assignments = panelAssignments
    if assignment != ButtonGroup.PanelAssignment.Unassigned { assignments[assignment.rawValue] = buttonGroup?.uuid }
    panelAssignments = assignments
  }

  /**
  buttonGroupForPanelAssignment:

  :param: assignment ButtonGroup.PanelAssignment

  :returns: ButtonGroup?
  */
  func buttonGroupForPanelAssignment(assignment: ButtonGroup.PanelAssignment) -> ButtonGroup? {
    var buttonGroup: ButtonGroup?
    if managedObjectContext != nil {
      if let uuid = panelAssignments[assignment.rawValue] {
        buttonGroup = ButtonGroup.existingObjectWithUUID(uuid, context: managedObjectContext!)
      }
    }
    return buttonGroup
  }


  /**
  updateWithData:

  :param: data [String:AnyObject]
  */
  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)

    if let moc = managedObjectContext {

      if let topBarHidden = data["top-bar-hidden"] as? NSNumber { self.topBarHidden = topBarHidden.boolValue }

      if let panels = data["panels"] as? [String:String] {
        for (key, uuid) in panels {
          if let buttonGroup = subelements.objectPassingTest({($0.0 as! RemoteElement).uuid == uuid}) as? ButtonGroup {
            let assignment = ButtonGroup.PanelAssignment(JSONValue: key)
            if assignment != ButtonGroup.PanelAssignment.Unassigned {
              setButtonGroup(buttonGroup, forPanelAssignment: assignment)
            }
          }
        }
      }

    }

  }

  /**
  JSONDictionary

  :returns: MSDictionary
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    let panels = MSDictionary()

    for (number, uuid) in panelAssignments {
      let assignment = ButtonGroup.PanelAssignment(rawValue: number.integerValue)
      if let commentedUUID = buttonGroupForPanelAssignment(assignment)?.commentedUUID {
        panels[assignment.JSONValue] = commentedUUID
      }
    }

    if panels.count > 0 { dictionary["panels"] = panels }

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }

}
